tt_metal/impl/device/device.cpp

```diff
@@ -3964,6 +3976,37 @@ std::vector<CoreCoord> Device::get_optimal_dram_bank_to_logical_worker_assignmen
     return this->optimal_dram_bank_to_logical_worker_assignment_;
 }
 
+HalProgrammableCoreType Device::get_programmable_core_type(CoreCoord virtual_core) const {
+
+    HalProgrammableCoreType programmable_core_type = HalProgrammableCoreType::TENSIX;
+    if (tt::Cluster::instance().is_ethernet_core(virtual_core, this->id_)) {
+        // Eth pcores have a different address, but only active ones.
+        CoreCoord logical_core = this->logical_core_from_ethernet_core(virtual_core);
+        if (this->is_active_ethernet_core(logical_core)) {
+            programmable_core_type = HalProgrammableCoreType::ACTIVE_ETH;
+        } else {
+            programmable_core_type = HalProgrammableCoreType::IDLE_ETH;
```

Comment:
Can we `return` and break the loop here, from the business logic perspective?

Btw, need to get rid of all of `this->`...